gsexp.el

```diff
@@ -0,0 +1,90 @@
+;;; gsexp.el --- GraphQl as S-expressions  -*- lexical-binding: t -*-
+
+;; Copyright (C) 2016-2019  Jonas Bernoulli
+
+;; Author: Jonas Bernoulli <jonas@bernoul.li>
+;; Homepage: https://github.com/magit/ghub
+;; Keywords: tools
+
+;; This file is not part of GNU Emacs.
+
+;; This file is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 3, or (at your option)
+;; any later version.
+
+;; This file is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; For a copy of the GPL see https://www.gnu.org/licenses/gpl.txt.
+
+;;; Commentary:
+
+;; This library encodes S-expressions as GraphQl documents.  Currently
+;; it is rather incomplete and only intended for use by `ghub-graphql'
+;; but I might fill the gaps and add documentation eventually.  I am
+;; not in a hurry though.
+
+;;; Code:
+
+(defun gsexp-encode (document)
+  (gsexp--pp (gsexp--encode-field document)))
+
+(defun gsexp--encode-field (field)
+  (cond
+   ((symbolp field)
+    (symbol-name field))
+   ((listp (car field))
+    (concat (format "%s: " (caar field))
+            (gsexp--encode-field (cons (cadar field)
+                                       (cdr field)))))
+   (t
+    (concat (symbol-name (pop field))
+            (and (vectorp (car field))
+                 (format " (\n%s)"
+                         (mapconcat #'gsexp--encode-argument
+                                    (append (pop field) nil)
+                                    ",\n")))
+            (and field
+                 (format " {\n%s\n}"
+                         (mapconcat #'gsexp--encode-field field "\n")))))))
+
+(cl-defun gsexp--encode-argument ((argument value))
+  (format "%s: %s" argument (gsexp--encode-value value)))
+
+(defun gsexp--encode-value (value)
+  (cond
+   ((numberp value)
+    (number-to-string value))
+   ((symbolp value) ; including variables, enums, booleans and null
+    (symbol-name value))
+   ((stringp value)
+    (prin1-to-string value))
```

Comment:
Do you need to handle `\n` and such here? (I had a post about this somewhere, but I can't seem to point my hand on it. There might be another variable to find here: https://emacs.stackexchange.com/questions/31252/can-i-tell-prin1-to-string-to-use-reader-macros#comment48033_31254)

Reply:
The variables of interest would be `print-escape-newlines` and `print-escape-control-characters`. And possibly the other `print-escape-*` variables (I don't know anything about graphQL, so I can't say what's needed).